### ISSUE TITLE
partial_ord_enum_variants_reordered Does not seem to allow inserting new variants in a non_exhaustive enum

### ISSUE THREAD

**Issue by cst0ny** [obi1kenobi/cargo-semver-checks#1244]
### Which lint or lints are the issue

Let me begin by saying that this `partial_ord_enum_variants_reordered` lint is awesome 👏 

I manage a project that has an enum where we add variants periodically. To make propagating this in our ecosystem easier, we use `#[non_exhaustive]` which forces our code to ignore "unknown-to-them" variants.

When inserting new variants in the enum, the `partial_ord_enum_variants_reordered` seems to FP and says that the ordering of the variant has changed. I don't think that is true in a `#[non_exhaustive]` context since all the existing/unchanged variants of the enum still order amongst themselves the same which means any existing code using these variants will not break.

### Known issues that might be causing this

- [ ] Is the flagged item defined in another crate, or a re-export of such an item? ([#638](https://github.com/obi1kenobi/cargo-semver-checks/issues/638))

### Steps to reproduce the bug with the above code

Inserting a new variant in an enum like :
```rust
#[derive(PartialEq, PartialOrd)]
#[non_exhaustive]
pub MyEnum {
    A,
    B,
    Z,
}
```

### Actual Behaviour

Adding a variant to the enum will trigger the lint
```rust
#[derive(PartialEq, PartialOrd)]
#[non_exhaustive]
pub MyEnum {
    A,
    B,
    // This is a new variant. All other variants still order the same relative to each other
    T,
    Z,
}
```

### Expected Behaviour

The rule should not trigger on `#[non_exhaustive]` enums when only new variants are added

### Verbose Lint Output

N/A

### Generated System Information

N/A

**Comment by Frank-III** [comment]
Thank you for raising this issue, good catch! We are currently checking if an enum variant position changed(`partial_ord_struct_fields_reordered` and `partial_ord_enum_struct_variant_fields_reordered` should have this issue too), and it seems we are not able to fix it except by restricting this lint more 🤔

**Comment by obi1kenobi** [comment]
Agreed — sorry for the inconvenience!

I think fixing this _currently_ would require making the lint a lot worse: we could look at all pairs of variants and lint on any that switched places, which will then trigger a zillion times if you move the last variant to first place.

But I think we might be able to fix it differently in a somewhat near-future version, by upgrading to a (still-unreleased) version of Trustfall that supports `@transform` on properties. If we can do arithmetic across fields, then _I think_ a variant has been reordered if: `current_position - count(added variants at lower positions) + count(removed variants that used to be at lower positions) != previous_position`.

Can folks check my logic here? In the opening example, for `Z`, `current_position = 4` (1-based), and one variant got newly added (`T`), and since `previous_position = 3` we get `3 - 1 + 0 = 3` so no breakage.

It would be good to make a large test corpus here for all the possible ways this could go wrong for both lints, so we can de-risk this approach and consequently inform us whether to try to ship that Trustfall feature sooner rather than later.

**Comment by stanislav-tkach** [comment]
I also have this issue. I understand that the fix isn't trivial, but is there a way to suppress this error for one enum specifically?

**Comment by obi1kenobi** [comment]
Unfortunately, not yet: https://github.com/obi1kenobi/cargo-semver-checks/issues/1491

The `cargo-semver-checks` project [funded a $1000 grant](https://bsky.app/profile/predr.ag/post/3m6ubedoonc2n) to the person who (among many other things) is working on the prerequisites for making that possible from the `rustc` side. If the ability to suppress lints on a per-item basis is important to you, please consider [funding `cargo-semver-checks`](https://github.com/sponsors/obi1kenobi/)!

In the meantime, your best bet is probably to [disable the lint as a whole](https://github.com/obi1kenobi/cargo-semver-checks?tab=readme-ov-file#configuration).
